### ISSUE TITLE
tf 2 vs tf 1 - different behaviour: model.evaluate(..., batch_size=...)

### ISSUE THREAD

**Issue by aciobanusebi** [tensorflow/tensorflow#38165]
Consider the following code (you can run it [here](https://colab.research.google.com/drive/1H04RM9TsFpQf8_8ZbExMkW-jWLzMi7eL) on colab):

```
import tensorflow as tf
from tensorflow import keras
import numpy as np

x_train = np.array([[1],[20],[3]],
                   dtype=np.float32)

y_train = np.array([[0.1],[0.2],[0.7]],
                   dtype=np.float32)

model = keras.Sequential([
                          keras.layers.Dense(1,input_shape=[1], activation='sigmoid',kernel_initializer='ones')
])
model.compile(
    loss=keras.losses.binary_crossentropy,
    optimizer=keras.optimizers.Adam(),
    metrics=['binary_crossentropy']
)

x_train.shape
model.evaluate(x_train,y_train,batch_size=3)
model.evaluate(x_train,y_train,batch_size=2)
model.evaluate(x_train[:2],y_train[:2],batch_size=2)
model.evaluate(x_train[2:],y_train[2:],batch_size=2)
```

In TF 2: it gives:

> **1/1** [==============================] - 0s 2ms/step - loss: 6.0539 - binary_crossentropy: 6.0539
**2/2** [==============================] - 0s 1ms/step - loss: **4.7776** - binary_crossentropy: 6.0539
**1/1** [==============================] - 0s 1ms/step - loss: 8.6066 - binary_crossentropy: 8.6066
1/1 [==============================] - 0s 1ms/step - loss: 0.9486 - binary_crossentropy: 0.9486
[0.9485874176025391, 0.9485874176025391]

In TF 1: it gives:

> **3/3** [==============================] - 0s 4ms/sample - loss: 6.0539 - binary_crossentropy: 6.0539
**3/3** [==============================] - 0s 2ms/sample - loss: **6.0539** - binary_crossentropy: 6.0539
**2/2** [==============================] - 0s 321us/sample - loss: 8.6066 - binary_crossentropy: 8.6066
1/1 [==============================] - 0s 713us/sample - loss: 0.9486 - binary_crossentropy: 0.9486
[0.9485874176025391, 0.9485874]

1. So, in tf 1 the evaluation is computed correctly irrespective of the batch_size, but in tf 2 it is not the case. Why?

2. Why the loss and metric are not the same: '2/2 [==============================] - 0s 1ms/step - loss: **4.7776** - binary_crossentropy: 6.0539' ?

**Comment by Saduf2019** [comment]
please find the gist for [tf nightly](https://colab.sandbox.google.com/gist/Saduf2019/107819b6a366ec2054ba831e49805406/2tf.ipynb) and [tf 1.x](https://colab.sandbox.google.com/gist/Saduf2019/ac58dbfe93c0dc4c8d9dc96ff9d313ee/1.ipynb), could you please let us know what is not computed correctly.

**Comment by aciobanusebi** [comment]
Yes. As far as I see the outputs are the same as in my first post. The difference I found is in bold: 4.7776  vs 6.0539.

**Comment by anneserious** [comment]
I have a similar isseu and would really be interested in the solution.
The validation loss should in general be independent of the batch size/order of data within batches, as long as it is evaluated on the entire dataset, right?

**Comment by bentyeh** [comment]
Please take a look at issue #38596, linked above. I believe this is the underlying issue for why the Keras `model.evaluate()` and `model.fit()` methods produce different values for different `batch_size` argument values.

**Comment by jvishnuvardhan** [comment]
@bentyeh This was resolved in recent tf-nightly. It will be available in stable TF2.2 in near future. [Here](https://colab.research.google.com/gist/jvishnuvardhan/8bbb847311468302c9ae99dd56e29590/untitled131.ipynb) is the gist for your reference. Thanks!

I am closing this issue as this was resolved. Please feel free to reopen if the issue persists again. Thanks!
